src/macros-by-example.adoc

```diff
@@ -0,0 +1,820 @@
+neologism will prove useful, you’ll see.
+
+The special-case E-expression `(:)` is synonymous with `(:void)`
+and exists solely as a more succinct argument expression for voidable parameters.
+
+
+=== Other Cardinalities
+
+For a language that’s thoroughly stream-oriented, rest-parameters don’t offer much flexibility
+with respect to cardinality.  The `*\...*` modifier can only be the final parameter, meaning that
+only that parameter can be invoked with more than one subexpression and produce more than one
+value.  So far other parameters must have exactly-one value, which is fairly limiting.
+
+In fact, the macro system provides four distinct cardinality constraints: exactly-one,
+zero-or-one, zero-or-more, and one-or-more.  They each have modifier symbols that can be used in
+signatures, namely `*!*`, `*?*`, `***`, and `*+*`. The former is the default, so the signatures
+`[arg]`, `[(*any* arg)]`, and `[(*any!* arg)]` are all equivalent.
+
+
+==== zero-or-one
+
+The zero-or-one cardinality, denoted in signatures by the modifier `*?*`, is much like the default
+exactly-one cardinality, except the parameter is voidable.  That is, it accepts an empty-stream
+argument as a way to denote an absent parameter.
+
+[{mrk}]
+----
+(*public* temperature
+  [(*decimal* degrees), (*symbol?* scale)]
+  {degrees: degrees, scale: scale})
+----
+
+Since the scale is voidable, we can pass it void:
+
+----
+(:temperature 96 F)    ⇒ {degrees:96, scale:F}
+(:temperature 283 (:)) ⇒ {degrees:283}
+----
+
+Note that the result’s `scale` field has disappeared because no value was provided.  It would be
+more useful to fill in a default value, and to do that we introduce another special form that can
+detect void:
+
+[{mrk}]
+----
+(*public* temperature
+  [(*decimal* degrees), (*symbol?* scale)]
+  {degrees: degrees, scale: (*if_void* scale (*literal* K) scale)})
+----
+----
+(:temperature 96 F)    ⇒ {degrees:96,  scale:F}
+(:temperature 283 (:)) ⇒ {degrees:283, scale:K}
+----
+
+The `*if_void*` form is if/then/else syntax testing stream emptiness. It has three sub-expressions,
+the first being a stream to check. If and only if that stream is void (_i.e._, it produces no
+values), the second sub-expression is expanded and its results are returned by the `*if_void*`
+expression. Otherwise, it produces at least one value, so the third sub-expression is expanded and
+returned.
+
+NOTE: Exactly one branch is expanded, because otherwise the void stream might be used in a context
+that requires a value, resulting in an errant expansion error.
+
+To refine things a bit further, trailing voidable arguments can be omitted entirely:
+
+----
+(:temperature 283) ⇒ {degrees:283, scale:K}
+----
+
+
+==== zero-or-more
+
+A zero-or-more `***` parameter is identical in meaning to our friend `*\...*` from the perspective
+of its template, but it can be used in any position, not just last place.
```

Comment:
*Minor*: Editorial suggestion.

Also, does the syntax for invoking differ between these two cardinalities?

```suggestion
A zero-or-more `***` parameter is identical in meaning to `*\...*` from the perspective
of its template, but it can be used in any position, not just last place.
```

Reply:
> Also, does the syntax for invoking differ between these two cardinalities?

Yeah, that's what the majority of this section discusses.  I think I'm missing your point...